### PR TITLE
Plugin restructure

### PR DESCRIPTION
**Pull request by italolelis**
## What does this PR do?

- Changed how our plugin architecture works. This will allow us in the future to move to go plugins natively (once they start supporting more platforms)
- Added more tests for plugins
- Added docs for plugins
- Added more examples on how to use Janus in many different ways

### CODE REVIEW

File: docs/quick_start/add_plugins.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# Add Plugins
+
+Any API Definition can make use of plugins. A plugin is a way to add functionalities to your definition, like rate limit, CORS, authentication, etc...
+
+In this tutorial we will add a [rate limit](/docs/plugins/rate_limit.md) plugin to our definition.
+
+## 1. Add the plugin
+
+Let's create a file called `rate_limit.json` with our new additions to the API definition. We will create a rate limit rula that we can only send `5 req/m`
```
Comment (rafaeljesus) [inline docs/quick_start/add_plugins.md]: typo `rula`

File: docs/quick_start/add_plugins.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      }
+    }
+  ]
+}
+```
+
+Now lets update our API definition:
+
+```sh
+http -v PUT localhost:8081/apis/my-endpoint "Authorization:Bearer yourToken" "Content-Type: application/json" < rate_limit.json
```
Comment (rafaeljesus) [inline docs/quick_start/add_plugins.md:30]: We need to expose `PATCH` method for changing proxies and use `PUT` for update the whole thing

File: docs/quick_start/add_plugins.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+You should see a few extra headers on the response of this request:
+
+```
+X-Ratelimit-Limit →5
+X-Ratelimit-Remaining →4
+X-Ratelimit-Reset →1498773715
+```
+
+This means the plugin is working properly. Now lets make the same requet 4 more times... On the fith time you should get:
```
Comment (rafaeljesus) [inline docs/quick_start/add_plugins.md]: typo `requet` and `fith`

File: examples/front-proxy-auth/apis/example.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        "strip_path" : false,
+        "append_path" : false,
+        "methods" : ["GET"]
+    },
+    "health_check": {
+        "url": "http://service1:8080/status",
+        "timeout": 3
+    },
+    "plugins": [
+                {
```
Comment (rafaeljesus) [inline examples/front-proxy-auth/apis/example.json]: Indentation of `{`

**Review by rafaeljesus — CHANGES_REQUESTED** [review]
Some typos, after that it's ready to go

**Review by rafaeljesus — APPROVED** [review]

**Comment by rafaeljesus** [comment]
👍 🚀 

[Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/159/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway) [Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/159/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway)

**Comment by italolelis** [comment]
👍 

[Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/159/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway) [Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/159/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway) [Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/159/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway)

**Comment by vgarvardt** [comment]
@italolelis +1 once more plz, I pushed one more tiny commit

**Comment by italolelis** [comment]
👍 

[Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/159/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway) [Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/159/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway) [Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/159/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway)

**Comment by italolelis** [comment]
Ow I just realized... It's my PR. @vgarvardt can you +1?

**Comment by vgarvardt** [comment]
👍 

[Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/159/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway) [Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/159/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway)

**Comment by rafaeljesus** [comment]
👍 

[Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/159/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway) [Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/159/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway)

**Pull request merged**